src/aposteriori/dnn/data_processing/tools.py

```diff
@@ -0,0 +1,158 @@
+
+    with h5py.File(frame_dataset, "r") as dataset_file:
+        flat_dataset_map = []
+        # Create flattened dataset structure:
+        for pdb_code in dataset_file:
+            for chain_id in dataset_file[pdb_code].keys():
+                for residue_id in dataset_file[pdb_code][chain_id].keys():
+                    residue_label = dataset_file[pdb_code][chain_id][
+                        str(residue_id)
+                    ].attrs["label"]
+
+                    if residue_label in standard_residues:
+                        pass
+                    # If uncommon, attempt conversion of label
+                    elif residue_label in UNCOMMON_RESIDUE_DICT.keys():
+                        print(f"{residue_label} is not a standard residue.")
+                        # Convert residue to common residue
+                        residue_label = UNCOMMON_RESIDUE_DICT[residue_label]
+                        print(f"Residue converted to {residue_label}.")
+                    else:
+                        print("Warning message")
+                        assert (
+                            residue_label in standard_residues
+                        ), f"Expected natural amino acid, but got {residue_label}."
+
+                    flat_dataset_map.append(
+                        (pdb_code, chain_id, residue_id, residue_label)
+                    )
+
+    return flat_dataset_map
+
+
+def balance_dataset(flat_dataset_map: t.List[t.Tuple]):
+    """
+    Balances the dataset by undersampling the least present residue.
+
+    Parameters
+    ----------
+    flat_dataset_map: t.List[t.Tuple]
+        List of tuples with the order
+        [... (pdb_code, chain_id, residue_id,  residue_label) ...]
+
+    Returns
+    -------
+    balanced_dataset_map: t.List[t.Tuple]
+        Balanced list of tuples with the order
+        [... (pdb_code, chain_id, residue_id,  residue_label) ...].
+        This is balanced by undersampling.
+    """
+    # Balancing the dataset:
+    shuffle(flat_dataset_map)
+    standard_residues = sorted(list(standard_amino_acids.values()))
+    # Count all residues in dataset:
+    all_residues_in_dataset = [res[-1] for res in flat_dataset_map]
+    # Count all residues and calculate the maximum number of residue per class:
+    counted_residue_in_dataset = Counter(all_residues_in_dataset)
+    # Count how many residues to extract per class:
+    max_res_num = counted_residue_in_dataset[
+        min(counted_residue_in_dataset, key=counted_residue_in_dataset.get)
+    ]
+    balanced_dataset_map = []
+    # Sort dataset by residue:
+    flat_dataset_map = sorted(flat_dataset_map, key=itemgetter(3))
+    for residue in standard_residues:
+        assert (
+                residue == flat_dataset_map[0][-1]
+        ), f"Expected {residue} residue but got {flat_dataset_map[0][-1]}"
+        # Append relevant residue:
+        balanced_dataset_map += flat_dataset_map[:max_res_num]
```

Comment:
I don't understand this, I thought that `flat_dataset_map` is just a list of tuples, how are making sure that the slice only contains one residue type?

Reply:
You are right, this is rather me being smart. I am unsure whether there's a better way to do this but:

 If you look at line 144 we sort by the type of residue:

`flat_dataset_map = sorted(flat_dataset_map, key=itemgetter(3))`

This means that there will always be residues of the same type close to each other. Now, we know that the max number of residues allowed (which is the **least** number of residue in any residue class, usually cysteine), so when we do:

`flat_dataset_map[:max_res_num]`

We select the residues required. For most classes, `max_res_num` < `res_num`. eg. n. of glutamates > n. of cysteines, so therefore we will have some leftovers. We remove the leftovers by using the counter and removing all the residues of that specific class. 

This is really convoluted, I'm thinking it should probably be better commented. Let me know if my explanation was clear, I can give it a second try :)

Reply:
I don't doubt your implementation, I'm sure it works, but I think partitioning the residues into a dict and then sampling from that would read better and would probably be easier to extend later on.

Reply:
yes you're absolutely right, a dictionary would be much simpler!